### DISCUSSION TITLE
Messages and laps

### DISCUSSION THREAD

**Discussion by McHouse** [theOehrly/Fast-F1#406] [Q&A] [answered]
Newbie problems again (I guess)

Is there a way to establish some reference between the messages of a session and the race laps in which they occur?

Thanks in advance.

**Comment by theOehrly** [comment]
> Newbie problems again (I guess)
> 
> Is there a way to establish some reference between the messages of a session and the race laps in which they occur?
> 
> 
> Thanks in advance.

The UTC timestamp of the message and the `LapStartTime` and `Time` of the laps dataframe have the same timebase (+- a few hundred milliseconds offset, maybe). So you can reference it to that.

**Reply by McHouse** [reply]
That's what I thought, but it turns out...
`GPname="MONACO"
session = fastf1.get_session(2023, 6, 'R')
session.load(telemetry=True, weather=True, messages=True)
laps = session.laps
Results = session.results
Messajes=session.race_control_messages
Weather = laps.get_weather_data()`

delivery
laps
	Time
0	0 days 01:03:27.458000
	LapStartTime
0	0 days 01:02:02.950000
	LapStartDate
0	2023-05-28 13:03:03.927

Messages
	Time
0	2023-05-28 12:05:10


I do not see it :(

**Reply by theOehrly** [reply]
Hi, sorry for the somewhat late response. You're right, of course, I remembered that wrong. 
The `Time` of the race control messages corresponds with the `Date` and `LapStartDate` of the lap timing data.

I have to say that this is a rather badly named to be fair. I might fix that in future releases, although that's somewhat difficult because it's a breaking change.

**Reply by McHouse** [reply]
Don't worry, no problem, I'm even surprised that you have time to review emails, questions, make corrections to the code, improvements, documentation... etc.
Thanks but ...
I still don't see a match.
Perhaps taking concrete facts, such as the closing of the pit lane prior to the start, but it seems to me a bad approximation.
even other more exact references, such as the appearance of any yellow flag (which would give a valid reference), even that may be absent in the anodyne GP...
And once again, THANK YOU.
(To you first, and then to other users who also help to clarify concepts and problems)

**Reply by McHouse** [reply]
Perhaps some kind collaborator can shed some light on this matter... Thank you.

**Comment by theOehrly** [comment, marked as answer]
Maybe I'm misunderstanding your question. Or you're confused by the data, which, to be fair, can be confusing 😅

Taking a look at the 2023 Monaco Race specifically, for example:

```
print(session.session_start_time)
1:02:02.950000
```
This tells you that the session only starts a bit more than 1 hour into the data, or at

```
print(session.t0_date + session.session_start_time)
2023-05-28 13:03:03.927000
```

Having a look at the race control messages, you can see the race starts at 13:03:04 which fits perfectly:

![grafik](https://github.com/theOehrly/Fast-F1/assets/23384863/7b51f32f-e53d-45f4-ab77-fe2a8ec46932)

Here is the session status data for comparison as well:

![grafik](https://github.com/theOehrly/Fast-F1/assets/23384863/a673239a-abbe-45a3-bdaf-772bd74727b6)


And here is Verstappen's timing data (partially). You can see that the first lap starts at 2023-05-28 13:03:03.927 again.

```
print(session.laps.pick_driver('VER').loc[:, ('LapTime', 'LapStartDate')])
                  LapTime            LapStartDate
0  0 days 00:01:24.238000 2023-05-28 13:03:03.927
1  0 days 00:01:19.367000 2023-05-28 13:04:28.435
2  0 days 00:01:19.074000 2023-05-28 13:05:47.802
3  0 days 00:01:18.129000 2023-05-28 13:07:06.876
4  0 days 00:01:18.019000 2023-05-28 13:08:25.005
..                    ...                     ...
73 0 days 00:01:26.864000 2023-05-28 14:44:47.818
74 0 days 00:01:25.518000 2023-05-28 14:46:14.682
75 0 days 00:01:25.316000 2023-05-28 14:47:40.200
76 0 days 00:01:25.460000 2023-05-28 14:49:05.516
77 0 days 00:01:25.201000 2023-05-28 14:50:30.976
[78 rows x 2 columns]
```

Does this answer your question now?

**Reply by McHouse** [reply]
Perfectly. You have made everything clear to me, as transparent as the interstellar void.
Really, my problem was in 'print(session.t0_date + session.session_start_time)' I didn't see the need for that sum.
Between my inexperience, the mess between the V2s (that I didn't get to control well), the new V3s and what is still messy for me (times and dates), and the changes with ergast...
I think I'm going to need "some time" to clarify myself better... hahahahaha.
Thank you very much theOehrly.
(you are earning coffees... be careful with sleep... :D )
